File: LeetCode/30.Substring_with_Concatenation_of_All_Words/Mario/30.Substring_with_Concatenation_of_All_Words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        Map<String, Integer> counter = new HashMap<>();
        for (String word : words) {
            counter.put(word, counter.getOrDefault(word, 0) + 1);
        }

        List<Integer> res = new ArrayList<>();

        int wordLen = words[0].length();
        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i, count = 0;

            Map<String, Integer> cur = new HashMap<>();

            while (right + wordLen <= s.length()) {
                String rWord = s.substring(right, right + wordLen);
                right += wordLen;

                if (counter.containsKey(rWord)) {
                    cur.put(rWord, cur.getOrDefault(rWord, 0) + 1);
                    count += 1;
                    while (cur.get(rWord) > counter.get(rWord)) {
                        String lword = s.substring(left, left + wordLen);
                        left += wordLen;
                        cur.put(lword, cur.get(lword) - 1);
                        count -= 1;
                    }

                    if (count == words.length) {
                        res.add(left);
                    }
                } else {
                    cur.clear();
                    left = right;
                    count = 0;
                }
            }
        }
        return res;
    }
}

File: LeetCode/30.Substring_with_Concatenation_of_All_Words/Mario/30.Substring_with_Concatenation_of_All_Words.py
#!/usr/bin/env python3


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        n, m, k = len(s), len(words), len(words[0])
        counter = Counter(words)

        res = []
        for i in range(k):
            # sliding window
            l = r = i
            cur = Counter()
            count = 0  # increment the word count in cur window

            while r + k <= n:
                w = s[r : r + k]
                r += k

                if w in counter:
                    cur[w] += 1
                    if cur[w] <= counter[w]:
                        count += 1

                    while cur[w] > counter[w]:
                        l_word = s[l : l + k]
                        cur[l_word] -= 1
                        if cur[l_word] < counter[l_word]:
                            count -= 1
                        l += k

                    if count == m:
                        res.append(l)
                        cur[s[l : l + k]] -= 1
                        count -= 1
                        l += k
                else:
                    cur.clear()
                    l = r
                    count = 0

        return res
